### ISSUE TITLE
Are you sure weather your sync-bn module works？

### ISSUE THREAD

**Issue by StevenGrove** [zhanghang1989/PyTorch-Encoding#53]
I tried one validation code, but it seems not works:

`
import torch
from torch.autograd import Variable
from torch.nn.parallel.data_parallel import DataParallel
from encoding.nn import BatchNorm2d

net = BatchNorm2d(128).cuda()import torch
from torch.autograd import Variable
from torch.nn.parallel.data_parallel import DataParallel
from encoding.nn import BatchNorm2d

net = BatchNorm2d(128).cuda()
net_sync = DataParallel(net, device_ids=list(range(8)))

while True:
    x = Variable(torch.randn(32, 128, 16, 16).cuda(), requires_grad=True)
    r = net(x)
    r_sync = net_sync(x)
    print(r.ne(r_sync).long().sum())
    print(torch.abs(r - r_sync).mean())
    loss = x.sum()
    loss.backward()
net_sync = DataParallel(net, device_ids=list(range(8)))

while True:
    x = Variable(torch.randn(32, 128, 16, 16).cuda(), requires_grad=True)
    r = net(x)
    r_sync = net_sync(x)
    print(r.ne(r_sync).long().sum())
    print(torch.abs(r - r_sync).mean())
    loss = x.sum()
    loss.backward()
`

**Results:**
tensor(1.0486e+06, device='cuda:0')
tensor(1.00000e-02 *
       1.3687, device='cuda:0')
tensor(1.0486e+06, device='cuda:0')
tensor(1.00000e-02 *
       1.3746, device='cuda:0')
tensor(1.0486e+06, device='cuda:0')
tensor(1.00000e-02 *
       1.3329, device='cuda:0')
tensor(1.0486e+06, device='cuda:0')
tensor(1.00000e-02 *
       1.3601, device='cuda:0')
tensor(1.0486e+06, device='cuda:0')
tensor(1.00000e-02 *
       1.4027, device='cuda:0')
tensor(1.0486e+06, device='cuda:0')
tensor(1.00000e-02 *
       1.2757, device='cuda:0')
...

**Comment by AssassinCroc** [comment]
Yep, there is definitely some issue in the training side of the code. Inference works for me.

**Comment by AssassinCroc** [comment]
Going to give this a try today. Awesome work!! Will keep you apprised!
